---
title: "MitoSignatures"
author: "Isabel Duarte"
date: "2023-06-05"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warnings=FALSE, cache = TRUE)

library(here)
library(tidyverse)
library(dplyr)
library(biomaRt)
```


##### Compare Gencode annotation files | Gencode v26 and v36

Compare by gene name and genomic coordinates 


```{r compare_gtf, warning=FALSE}

# NOTE | The annotation files were download manually from:
# https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_36/gencode.v36.annotation.gtf.gz
# https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_36/gencode.v36.annotation.gff3.gz

# https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_26/gencode.v26.annotation.gtf.gz
# https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_26/gencode.v26.annotation.gff3.gz

####

# Path to the GTF files
gtf_file_v26 <- here("data/annotation_files/gencode.v26.annotation.gtf")
gtf_file_v36 <- here("data/annotation_files/gencode.v36.annotation.gtf")

# Read the GTF files
gtf_v26 <- read_delim(gtf_file_v26, 
                      delim = "\t", escape_double = FALSE, 
                      col_names = c("chromosome_name",
                                    "annotation_source",
                                    "feature_type",
                                    "genomic_start",
                                    "genomic_end",
                                    "score",
                                    "genomic_strand",
                                    "genomic_phase",
                                    "attributes"), 
                      trim_ws = TRUE, comment = "#")

gtf_v36 <- read_delim(gtf_file_v36, 
                      delim = "\t", escape_double = FALSE, 
                      col_names = c("chromosome_name",
                                    "annotation_source",
                                    "feature_type",
                                    "genomic_start",
                                    "genomic_end",
                                    "score",
                                    "genomic_strand",
                                    "genomic_phase",
                                    "attributes"), 
                      trim_ws = TRUE, comment = "#")

# Filter only the gene feature
gtf_v26_genes <- filter (gtf_v26,
                         feature_type == "gene") 

gtf_v36_genes <- filter (gtf_v36,
                         feature_type == "gene") 


# Intersect both gene annotation sets
gtf_common_genes <- inner_join(
  gtf_v26_genes,
  gtf_v36_genes,
  by = c("chromosome_name",
         "annotation_source",
         "feature_type",
         "genomic_start",
         "genomic_end",
         "score",
         "genomic_strand",
         "genomic_phase")
)

# Get the ENSEMBL gene id from the attributes column
gtf_common_genes$gene_id_v26 <- sub('.*gene_id\\s+"(ENSG\\d+(\\.\\d+)?(\\S+)?)".*', '\\1', gtf_common_genes$attributes.x)
gtf_common_genes$gene_id_v36 <- sub('.*gene_id\\s+"(ENSG\\d+(\\.\\d+)?(\\S+)?)".*', '\\1', gtf_common_genes$attributes.y)

# Compare the two gene_id columns
unique_genes_v26 <- gtf_common_genes$gene_id_v26[!gtf_common_genes$gene_id_v26 %in% gtf_common_genes$gene_id_v36]
unique_genes_v36 <- gtf_common_genes$gene_id_v36[!gtf_common_genes$gene_id_v36 %in% gtf_common_genes$gene_id_v26]

# Test if the only difference is in gene version number (numbers after the stable identifier and the .)
genes_v26_not_36 <- unique_genes_v26[!(gsub("\\.\\d+","", unique_genes_v26) %in% gsub("\\.\\d+","", unique_genes_v36))]
genes_v36_not_26 <- unique_genes_v36[!(gsub("\\.\\d+","", unique_genes_v36) %in% gsub("\\.\\d+","", unique_genes_v26))]

# Identity of the different genes (using the stable identifier only)
genes_v36_not_26[order(genes_v36_not_26)]
genes_v26_not_36[order(genes_v26_not_36)]

```


# Creating a dataframe with all of the different genes for v36 showing their symbol and name

```{r dataframe for genes_v36, warning=FALSE}

ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")

# Creating dataframe for genes_v36_not_26 with gene symbol and name
gene_ids_v36 <- strsplit(genes_v36_not_26[order(genes_v36_not_26)], " ")

ensembl_id_without_version_v36 <- sub("\\..*", "", gene_ids_v36)

# Fetch gene symbols and gene names
gene_info_v36 <- getBM(
  attributes = c("ensembl_gene_id", "external_gene_name", "description"),
  filters = "ensembl_gene_id",
  values = ensembl_id_without_version_v36,
  mart = ensembl
)

# Create a dataframe
#### ITS REMOVING ONE OF THE GENES (SHOULD BE 52 NOT 51)
gene_v36_dataframe <- data.frame(row.names = gene_info_v36$ensembl_gene_id, 
                                 Symbol = gene_info_v36$external_gene_name,
                                 Name = gene_info_v36$description)

# Save the id not present in the biomart database
ensembl_id_without_version_v36_missing <- ensembl_id_without_version_v36[!(ensembl_id_without_version_v36 %in% rownames(gene_v36_dataframe))]

# View the dataframe
print(gene_v36_dataframe)

# Export the dataframe to a csv file
write.csv(gene_v36_dataframe, here("analysis_output/gene_v36_dataframe.csv"), row.names=TRUE)

```


# Creating a dataframe with all of the different genes for v26 showing their symbol and name

```{r dataframe for genes_v26, warning=FALSE}

ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")

# Creating dataframe for genes_v26_not_36 with gene symbol and name
gene_ids_v26 <- strsplit(genes_v26_not_36[order(genes_v26_not_36)], " ")

ensembl_id_without_version_v26 <- sub("\\..*", "", gene_ids_v26)

# Fetch gene symbols and gene names
gene_info_v26 <- getBM(
  attributes = c("ensembl_gene_id", "external_gene_name", "description"),
  filters = "ensembl_gene_id",
  values = ensembl_id_without_version_v26,
  mart = ensembl
)

# Create a dataframe
gene_v26_dataframe <- data.frame(row.names = gene_info_v26$ensembl_gene_id, 
                                 Symbol = gene_info_v26$external_gene_name,
                                 Name = gene_info_v26$description)

rownames(gene_v26_dataframe) <- gene_info_v26$ensembl_gene_id

# Save the id not present in the biomart database
ensembl_id_without_version_v26_missing <- ensembl_id_without_version_v26[!(ensembl_id_without_version_v26 %in% rownames(gene_v26_dataframe))] 

# View the dataframe
print(gene_v26_dataframe)

# Export the dataframe to a csv file
write.csv(gene_v26_dataframe, here("analysis_output/gene_v26_dataframe.csv"), row.names=TRUE)

```


## PCA from TCGA GTEx combined counts

```{r pca_analysis, warning=FALSE}

# Load data
combined_pca_result <- readRDS(here("data/processed/combined_pca_result.RDS"))
combined_metadata_tcga_gtex <- readRDS(here("data/processed/combined_metadata_tcga_gtex.RDS"))

# Create a dataframe with PC scores, patient id, and study
data.frame(PC1 = combined_pca_result$x[, 1],
           PC2 = combined_pca_result$x[, 2]) %>% 
  rownames_to_column(var="sample_id") %>%
  mutate(study = case_when(grepl("^TCGA", sample_id) ~ "TCGA",
                           grepl("^GTEX", sample_id) ~ "GTEX",
                           TRUE ~ NA_character_)) -> pca_scores_df

# Making left join of pca_results_df with combined_metadata_tcga_gtex through the column "sample_id"
result_df <- left_join(pca_scores_df, combined_metadata_tcga_gtex, by = "sample_id")

# Add the "organ" column to pca_scores_df
new_pca_scores_df <- mutate(pca_scores_df, organ = result_df$organ)

# How to calculate the percentage of variance explained 
variance_explained <- (combined_pca_result$sdev^2) / sum(combined_pca_result$sdev^2) * 100

# Look at the results: PC1 vs PC2 scatter plot
pca_plot_study <- ggplot(new_pca_scores_df, aes(x = PC1, y = PC2, color=study)) +
  geom_point(alpha=0.5) +
  geom_text(data = filter(new_pca_scores_df, PC1>= 1e+07), aes(label = organ), vjust = 1.5, size = 3) +
  xlab("PC1 (78.054 %)") +
  ylab("PC2 (7.268 %)") +
  ggtitle("PCA Plot | Combined TCGA and GTEx data")

print(pca_plot_study)

# Look at the results: PC1 vs PC2 scatter plot
pca_plot_organ <- ggplot(new_pca_scores_df, aes(x = PC1, y = PC2, color=organ)) +
  geom_point(aes(shape = study), alpha=0.5, size = 2) +
  geom_text(data = filter(new_pca_scores_df, PC1>= 1e+07), aes(label = organ), vjust = 1.5, size = 3) +
  xlab("PC1 (78.054 %)") +
  ylab("PC2 (7.268 %)") +
  ggtitle("PCA Plot | Combined TCGA and GTEx data")

print(pca_plot_organ)

```


```{r saving pca_analysis, warning=FALSE}

## Saving pca plot
ggsave(file=here("analysis_output/pca_plot_study.png"),
    plot = pca_plot_study, 
    device="png", width = 21.15, height = 15, units = "cm" 
    )

## Saving pca plot
ggsave(file=here("analysis_output/pca_plot_organ.png"),
    plot = pca_plot_organ, 
    device="png", width = 21.15, height = 15, units = "cm" 
    )
```
